### PR TITLE
Add withFunctionsTriggersDisabled method to rules-unit-testing

### PR DESCRIPTION
**Pull request by samtstern**
### Discussion

Note to reviewer: this makes two changes to dev dependencies:
  * Updated to `firebase-tools` `8.13.0` which has https://github.com/firebase/firebase-tools/pull/2706
  * Added `firebase-functions` because in order to properly test this we needed the functions emulator to run during the tests (even with no functions).  The functions emulator refuses to start up if it can't find `firebase-functions` in the developer's `node_modules`.

### Testing

  * Added unit tests

### API Changes

Googlers: See API proposal [here](https://docs.google.com/document/d/1mBOYZKJv6-S1Rfzb25FVkv3yJSjrXeFykeCclQZwGiM/edit?usp=sharing)

### CODE REVIEW

File: packages/rules-unit-testing/src/api/index.ts [hunk 1]
```diff
@@ -317,33 +322,25 @@ export function loadDatabaseRules(
...
-        if (err) {
-          reject(err);
-        } else if (resp.statusCode !== 200) {
-          reject(JSON.parse(body).error);
-        } else {
-          resolve();
-        }
-      }
-    );
+  const resp = await requestPromise({
```
Comment (samtstern) [inline packages/rules-unit-testing/src/api/index.ts:325]: I also DRY-ed up a lot of the `request` boillerplate in this file.

File: packages/rules-unit-testing/src/api/index.ts [hunk 2]
```diff
@@ -354,64 +351,96 @@ export function loadFirestoreRules(
...
+    console.warn(
+      `${HUB_HOST_ENV} is not set, assuming the Emulator hub is running at ${HUB_HOST_DEFAULT}`
     );
+    hubHost = HUB_HOST_DEFAULT;
+  }
+
+  // Disable background triggers
+  const disableRes = await requestPromise({
+    method: 'PUT',
+    uri: `http://${hubHost}/functions/disableBackgroundTriggers/`
```
Comment (yuchenshi) [inline packages/rules-unit-testing/src/api/index.ts:420]:
```suggestion
    uri: `http://${hubHost}/functions/disableBackgroundTriggers`
```

File: packages/rules-unit-testing/src/api/index.ts [hunk 3]
```diff
@@ -354,64 +351,96 @@ export function loadFirestoreRules(
...
+    );
+  }
+
+  // Run the user's function
+  const result = await fn();
+
+  // Re-enable background triggers
+  const enableRes = await requestPromise({
+    method: 'PUT',
+    uri: `http://${hubHost}/functions/enableBackgroundTriggers/`
```
Comment (yuchenshi) [inline packages/rules-unit-testing/src/api/index.ts:434]:
```suggestion
    uri: `http://${hubHost}/functions/enableBackgroundTriggers`
```

File: packages/rules-unit-testing/src/api/index.ts [hunk 4]
```diff
@@ -354,64 +351,96 @@ export function loadFirestoreRules(
...
+    throw new Error(
+      `HTTP Error ${disableRes.statusCode} when disabling functions triggers, are you using the latest version of the Firebase CLI?`
+    );
+  }
+
+  // Run the user's function
+  const result = await fn();
+
+  // Re-enable background triggers
+  const enableRes = await requestPromise({
```
Comment (yuchenshi) [inline packages/rules-unit-testing/src/api/index.ts:432]: Use a `finally` block please to make sure this is executed even if `fn()` throws. Like:

`try { return await fn(); } finally { /* re-enable triggers */ }`

**Review by yuchenshi — APPROVED** [review]

File: packages/rules-unit-testing/src/api/index.ts (hunk 4 above)
Comment (samtstern) [reply packages/rules-unit-testing/src/api/index.ts]: Good call

**Review by yuchenshi — APPROVED** [review]

**Review by henrypbriffel — APPROVED** [review]

File: packages/rules-unit-testing/package.json [hunk 5]
```diff
@@ -28,7 +28,8 @@
     "@google-cloud/firestore": "4.4.0",
     "@types/request": "2.48.5",
     "firebase-admin": "9.2.0",
-    "firebase-tools": "8.12.1",
+    "firebase-tools": "8.13.0",
+    "firebase-functions": "3.11.0",
```
Comment (samtstern) [inline packages/rules-unit-testing/package.json:32]: See note in PR description about this new dev dependency

File: .changeset/lazy-elephants-suffer.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+'firebase': minor
```
Comment (Feiyang1) [inline .changeset/lazy-elephants-suffer.md:2]:
```suggestion
```
The version of `@firebase/rules-unit-testing` doesn't affect `firebase`

File: packages/rules-unit-testing/src/api/index.ts [hunk 7]
```diff
@@ -355,64 +352,98 @@ export function loadFirestoreRules(
...
+    throw new Error(JSON.parse(resp.body.error));
+  }
+}
+
+/**
+ * Run a setup function with background Cloud Functions triggers disabled. This can be used to
+ * import data into the Realtime Database or Cloud Firestore emulator without triggering locally
+ * emulated Cloud Functions.
+ *
+ * This method only works with Firebase CLI version 8.13.0 or higher.
```
Comment (Feiyang1) [inline packages/rules-unit-testing/src/api/index.ts:402]: Do you mind adding this comment to changelog to provide some details?

**Review by Feiyang1 — CHANGES_REQUESTED** [review]

File: packages/rules-unit-testing/src/api/index.ts (hunk 7 above)
Comment (samtstern) [reply packages/rules-unit-testing/src/api/index.ts:402]: Done, although I wasn't quite sure about the changelog format.  Most of the other messages seem to be machine generated?

Comment (Feiyang1) [reply packages/rules-unit-testing/src/api/index.ts:402]: Sorry, I wasn't clear. I meant adding it to the changeset. Yes, the changelog is generated from changesets.

Comment (samtstern) [reply packages/rules-unit-testing/src/api/index.ts:402]: Got it, fixed!

**Review by Feiyang1 — APPROVED** [review]

**Comment by Feiyang1** [comment]
Feel free to merge. I will fix the changeset warning in a different PR.

**Comment by saculbr** [comment]
This should work with Firebase triggers?


because I tried to call *firebase.clearFirestoreData* wrapped *with withFunctionTriggersDisabled* but it didn't work as I expected.

My enviroment
Linux / Ubuntu 20.04
firebase-tools 8.16.1
@firebase/rules-unit-testing 1.1.0
firebase-admin 9.4.0
firebase 8.0.1

Code used (I can create an separate issue with an proper repo if needed)
```
beforeEach(async () => {
    await firebase.withFunctionTriggersDisabled(() => firebase.clearFirestoreData({ projectId: APP_ID }));
});
```

Selection_020

**Comment by samtstern** [comment]
@saculbr you have a hanging promise. If you add 'return' before clearFirestoreData it should work. 

If not, please file a new issue!

**Comment by saculbr** [comment]
@samtstern I tried that too, but don't worked

**Comment by samtstern** [comment]
@saculbr you're totally right, this is broken.

**Pull request merged**